File: shared/go_libs/validation/password.go
package validation

import (
	"regexp"
	"strings"
	"unicode"
)

// PasswordOptions configures password validation requirements.
type PasswordOptions struct {
	MinLength        int
	MaxLength        int
	RequireUppercase bool
	RequireLowercase bool
	RequireDigit     bool
	RequireSpecial   bool
	SpecialChars     string
	DisallowSpaces   bool
}

// DefaultPasswordOptions returns default password options.
func DefaultPasswordOptions() PasswordOptions {
	return PasswordOptions{
		MinLength:        8,
		MaxLength:        128,
		RequireUppercase: true,
		RequireLowercase: true,
		RequireDigit:     true,
		RequireSpecial:   true,
		SpecialChars:     "!@#$%^&*()_+-=[]{}|;:,.<>?~`",
		DisallowSpaces:   true,
	}
}

// WeakPasswordOptions returns weak password requirements.
func WeakPasswordOptions() PasswordOptions {
	return PasswordOptions{
		MinLength:        6,
		MaxLength:        128,
		RequireUppercase: false,
		RequireLowercase: false,
		RequireDigit:     false,
		RequireSpecial:   false,
		SpecialChars:     "!@#$%^&*()_+-=[]{}|;:,.<>?~`",
		DisallowSpaces:   true,
	}
}

// ModeratePasswordOptions returns moderate password requirements.
func ModeratePasswordOptions() PasswordOptions {
	return PasswordOptions{
		MinLength:        8,
		MaxLength:        128,
		RequireUppercase: true,
		RequireLowercase: true,
		RequireDigit:     true,
		RequireSpecial:   false,
		SpecialChars:     "!@#$%^&*()_+-=[]{}|;:,.<>?~`",
		DisallowSpaces:   true,
	}
}

// StrongPasswordOptions returns strong password requirements.
func StrongPasswordOptions() PasswordOptions {
	return PasswordOptions{
		MinLength:        12,
		MaxLength:        128,
		RequireUppercase: true,
		RequireLowercase: true,
		RequireDigit:     true,
		RequireSpecial:   true,
		SpecialChars:     "!@#$%^&*()_+-=[]{}|;:,.<>?~`",
		DisallowSpaces:   true,
	}
}

// EnterprisePasswordOptions returns enterprise password requirements.
func EnterprisePasswordOptions() PasswordOptions {
	return PasswordOptions{
		MinLength:        16,
		MaxLength:        256,
		RequireUppercase: true,
		RequireLowercase: true,
		RequireDigit:     true,
		RequireSpecial:   true,
		SpecialChars:     "!@#$%^&*()_+-=[]{}|;:,.<>?~`",
		DisallowSpaces:   true,
	}
}

// StrongPassword validates password strength based on configurable requirements.
func StrongPassword(opts ...func(*passwordOpts)) Validator {
	o := &passwordOpts{
		options: DefaultPasswordOptions(),
	}
	for _, opt := range opts {
		opt(o)
	}

	return &StringValidator{
		validate: func(s string) ValidationResult {
			var errors []string
			po := o.options

			// Length checks
			if len(s) < po.MinLength {
				errors = append(errors, "password must be at least "+itoa(po.MinLength)+" characters")
			}

			if len(s) > po.MaxLength {
				errors = append(errors, "password must be at most "+itoa(po.MaxLength)+" characters")
			}

			// Space check
			if po.DisallowSpaces && strings.Contains(s, " ") {
				errors = append(errors, "password cannot contain spaces")
			}

			// Character type checks
			if po.RequireUppercase && !hasUppercase(s) {
				errors = append(errors, "password must contain at least one uppercase letter")
			}

			if po.RequireLowercase && !hasLowercase(s) {
				errors = append(errors, "password must contain at least one lowercase letter")
			}

			if po.RequireDigit && !hasDigit(s) {
				errors = append(errors, "password must contain at least one digit")
			}

			if po.RequireSpecial && !hasSpecial(s, po.SpecialChars) {
				errors = append(errors, "password must contain at least one special character")
			}

			if len(errors) > 0 {
				if o.errorMessage != "" {
					return Failure(o.errorMessage)
				}
				return Failure(strings.Join(errors, "; "))
			}

			return Success(s)
		},
	}
}

type passwordOpts struct {
	options      PasswordOptions
	errorMessage string
}

// WithPasswordOptions sets custom password options.
func WithPasswordOptions(opts PasswordOptions) func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.options = opts
	}
}

// WithMinPasswordLength sets minimum password length.
func WithMinPasswordLength(min int) func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.options.MinLength = min
	}
}

// WithMaxPasswordLength sets maximum password length.
func WithMaxPasswordLength(max int) func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.options.MaxLength = max
	}
}

// WithPasswordError sets a custom error message.
func WithPasswordError(msg string) func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.errorMessage = msg
	}
}

// WithoutUppercase removes uppercase requirement.
func WithoutUppercase() func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.options.RequireUppercase = false
	}
}

// WithoutLowercase removes lowercase requirement.
func WithoutLowercase() func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.options.RequireLowercase = false
	}
}

// WithoutDigit removes digit requirement.
func WithoutDigit() func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.options.RequireDigit = false
	}
}

// WithoutSpecial removes special character requirement.
func WithoutSpecial() func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.options.RequireSpecial = false
	}
}

// AllowSpaces allows spaces in passwords.
func AllowSpaces() func(*passwordOpts) {
	return func(o *passwordOpts) {
		o.options.DisallowSpaces = false
	}
}

// Helper functions

func hasUppercase(s string) bool {
	for _, r := range s {
		if unicode.IsUpper(r) {
			return true
		}
	}
	return false
}

func hasLowercase(s string) bool {
	for _, r := range s {
		if unicode.IsLower(r) {
			return true
		}
	}
	return false
}

func hasDigit(s string) bool {
	for _, r := range s {
		if unicode.IsDigit(r) {
			return true
		}
	}
	return false
}

func hasSpecial(s, specialChars string) bool {
	specialSet := make(map[rune]struct{}, len(specialChars))
	for _, r := range specialChars {
		specialSet[r] = struct{}{}
	}

	for _, r := range s {
		if _, ok := specialSet[r]; ok {
			return true
		}
	}
	return false
}

// itoa is a simple int to string conversion
func itoa(i int) string {
	if i == 0 {
		return "0"
	}

	var negative bool
	if i < 0 {
		negative = true
		i = -i
	}

	var digits []byte
	for i > 0 {
		digits = append([]byte{byte('0' + i%10)}, digits...)
		i /= 10
	}

	if negative {
		return "-" + string(digits)
	}
	return string(digits)
}

// PasswordStrengthScore calculates a password strength score (0-100).
func PasswordStrengthScore(password string, specialChars string) int {
	if specialChars == "" {
		specialChars = "!@#$%^&*()_+-=[]{}|;:,.<>?~`"
	}

	score := 0

	// Length contribution (up to 30 points)
	lengthScore := len(password) * 2
	if lengthScore > 30 {
		lengthScore = 30
	}
	score += lengthScore

	// Character variety (up to 40 points)
	variety := 0
	if hasLowercase(password) {
		variety++
	}
	if hasUppercase(password) {
		variety++
	}
	if hasDigit(password) {
		variety++
	}
	if hasSpecial(password, specialChars) {
		variety++
	}
	score += variety * 10

	// Unique character ratio (up to 20 points)
	if len(password) > 0 {
		unique := make(map[rune]struct{})
		for _, r := range password {
			unique[r] = struct{}{}
		}
		uniqueRatio := float64(len(unique)) / float64(len(password))
		score += int(uniqueRatio * 20)
	}

	// No common patterns bonus (up to 10 points)
	commonPatterns := []*regexp.Regexp{
		regexp.MustCompile(`^123`),
		regexp.MustCompile(`(?i)abc`),
		regexp.MustCompile(`(?i)qwerty`),
		regexp.MustCompile(`(?i)password`),
		regexp.MustCompile(`(.)\1{2,}`), // Repeated characters
	}

	hasCommon := false
	for _, pattern := range commonPatterns {
		if pattern.MatchString(password) {
			hasCommon = true
			break
		}
	}
	if !hasCommon {
		score += 10
	}

	if score > 100 {
		score = 100
	}

	return score
}

File: shared/node_libs/src/validation/password.ts
/**
 * Password validators for input validation.
 */

import { type ValidationResult, type Validator, success, failure } from './index.js';

export interface PasswordOptions {
  minLength?: number;
  maxLength?: number;
  requireUppercase?: boolean;
  requireLowercase?: boolean;
  requireDigit?: boolean;
  requireSpecial?: boolean;
  specialChars?: string;
  disallowSpaces?: boolean;
}

const DEFAULT_OPTIONS: Required<PasswordOptions> = {
  minLength: 8,
  maxLength: 128,
  requireUppercase: true,
  requireLowercase: true,
  requireDigit: true,
  requireSpecial: true,
  specialChars: '!@#$%^&*()_+-=[]{}|;:,.<>?~`',
  disallowSpaces: true,
};

/**
 * Returns weak password options (min 6 chars, no special requirements).
 */
export function weakPasswordOptions(): PasswordOptions {
  return {
    minLength: 6,
    requireUppercase: false,
    requireLowercase: false,
    requireDigit: false,
    requireSpecial: false,
  };
}

/**
 * Returns moderate password options (min 8 chars, letters and digits).
 */
export function moderatePasswordOptions(): PasswordOptions {
  return {
    minLength: 8,
    requireUppercase: true,
    requireLowercase: true,
    requireDigit: true,
    requireSpecial: false,
  };
}

/**
 * Returns strong password options (min 12 chars, all character types).
 */
export function strongPasswordOptions(): PasswordOptions {
  return {
    minLength: 12,
    requireUppercase: true,
    requireLowercase: true,
    requireDigit: true,
    requireSpecial: true,
  };
}

/**
 * Returns enterprise password options (min 16 chars, all character types).
 */
export function enterprisePasswordOptions(): PasswordOptions {
  return {
    minLength: 16,
    maxLength: 256,
    requireUppercase: true,
    requireLowercase: true,
    requireDigit: true,
    requireSpecial: true,
  };
}

export interface StrongPasswordValidatorOptions extends PasswordOptions {
  errorMessage?: string;
}

/**
 * Validates password strength based on configurable requirements.
 */
export function strongPassword(options: StrongPasswordValidatorOptions = {}): Validator<string, string> {
  const opts: Required<PasswordOptions> = { ...DEFAULT_OPTIONS, ...options };

  return (value: string): ValidationResult<string> => {
    if (typeof value !== 'string') {
      return failure('Password must be a string');
    }

    const errors: string[] = [];

    // Length checks
    if (value.length < opts.minLength) {
      errors.push(`Password must be at least ${opts.minLength} characters`);
    }

    if (value.length > opts.maxLength) {
      errors.push(`Password must be at most ${opts.maxLength} characters`);
    }

    // Space check
    if (opts.disallowSpaces && value.includes(' ')) {
      errors.push('Password cannot contain spaces');
    }

    // Character type checks
    if (opts.requireUppercase && !/[A-Z]/.test(value)) {
      errors.push('Password must contain at least one uppercase letter');
    }

    if (opts.requireLowercase && !/[a-z]/.test(value)) {
      errors.push('Password must contain at least one lowercase letter');
    }

    if (opts.requireDigit && !/\d/.test(value)) {
      errors.push('Password must contain at least one digit');
    }

    if (opts.requireSpecial) {
      const specialRegex = new RegExp(`[${escapeRegex(opts.specialChars)}]`);
      if (!specialRegex.test(value)) {
        errors.push('Password must contain at least one special character');
      }
    }

    if (errors.length > 0) {
      if (options.errorMessage) {
        return failure(options.errorMessage);
      }
      return failure(errors.join('; '));
    }

    return success(value);
  };
}

/**
 * Calculates a password strength score (0-100).
 */
export function passwordStrengthScore(
  password: string,
  specialChars: string = DEFAULT_OPTIONS.specialChars
): number {
  if (!password) return 0;

  let score = 0;

  // Length contribution (up to 30 points)
  score += Math.min(password.length * 2, 30);

  // Character variety (up to 40 points)
  let variety = 0;
  if (/[a-z]/.test(password)) variety++;
  if (/[A-Z]/.test(password)) variety++;
  if (/\d/.test(password)) variety++;
  const specialRegex = new RegExp(`[${escapeRegex(specialChars)}]`);
  if (specialRegex.test(password)) variety++;
  score += variety * 10;

  // Unique character ratio (up to 20 points)
  const unique = new Set(password).size;
  const uniqueRatio = unique / password.length;
  score += Math.floor(uniqueRatio * 20);

  // No common patterns bonus (up to 10 points)
  const commonPatterns = [
    /^123/,
    /abc/i,
    /qwerty/i,
    /password/i,
    /(.)\1{2,}/, // Repeated characters
  ];

  const hasCommon = commonPatterns.some(pattern => pattern.test(password));
  if (!hasCommon) {
    score += 10;
  }

  return Math.min(score, 100);
}

/**
 * Gets a human-readable strength label for a password score.
 */
export function passwordStrengthLabel(score: number): 'very weak' | 'weak' | 'fair' | 'strong' | 'very strong' {
  if (score < 20) return 'very weak';
  if (score < 40) return 'weak';
  if (score < 60) return 'fair';
  if (score < 80) return 'strong';
  return 'very strong';
}

// Helper to escape special regex characters
function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
